### ISSUE TITLE
Logging

### ISSUE THREAD

**Issue by darkl** [Code-Sharp/WampSharp#6]
There should be logs in the framework.

A logging framework (such as log4net) should be considered.

**Comment by darkl** [comment]
Being developed on logs branch using Castle ILogger interface. An interesting option is to use [LibLog](https://github.com/damianh/LibLog).

Please comment here about specific logs you would like to have from the library.

**Comment by shanielh** [comment]
Use LibLog :+1:

**Comment by fconrady** [comment]
In the case of Autobahn's Python, JavaScript and C++ clients I found the debug flag helpful. Something similar may be good for WampSharp, for example, being able to see the JSON that was serialized or deserialized.

**Comment by darkl** [comment]
Yeah, the serialized json is actually written to logs in the develop branch. You are welcome to try it via [MyGet feed](https://www.myget.org/feed/Packages/wampsharp) and add other recommendations.

**Comment by BrannonKing** [comment]
If you want to expose some Monitor event for logging, that might be handy. Trace calls (System.Diagnostics) would also work, but I generally prefer them less. I would not integrate some 3rd-party logging solution. Leave that to the application-level integration so that they can chose the right logging library for the application -- it's not a framework thing.

**Comment by RichiCoder1** [comment]
@darkl I actually ran into a perf issue because I was sending not-small json messages over Wamp and the sync logging was waiting to write out massive strings before continueing. I switched the logger to async, but I was wondering if was possible to either A) only write messages on a parse failure or B) but message writing behind it's own flag completely.

**Comment by darkl** [comment]
Option A is already possible, just set the log level of the logger
WampSharp.Newtonsoft.JTokenMessageParser to Error level.

Elad

**Comment by RichiCoder1** [comment]
There's no way to it specifically though, since it pulls the logger effectively out of thin error w/ LibLog. So I'd have to set the context of my whole application to something higher than debug.

**Comment by darkl** [comment]
I am not around a computer right now, but for example for log4net you need
to create a Logger.config with a ROOT logger set to debug and a logger with
the name I mentioned set to ERROR.

Then you need to call XmlConfigurator.Configure. This should work.

Elad

**Comment by darkl** [comment]
See here for the exact syntax.

https://logging.apache.org/log4net/release/manual/configuration.html

**Comment by RichiCoder1** [comment]
I utilize Serilog, but the issue is I'd like to keep the log level set to Debug and have Json messages not automatically be written, even at Debug level.

**Comment by darkl** [comment]
So manually change the log level of this specific logger. This is one of
the things logging frameworks are for.

**Comment by RichiCoder1** [comment]
Apparently Serilog added a source context minium level override [a few months back](https://nblumhardt.com/2016/07/serilog-2-minimumlevel-override/), so my point is indeed moot. Sorry about that.

**Comment by darkl** [comment]
Great!

**Comment by llehn** [comment]
LibLog has absolutely horrifying performance. It should not be used in a library/framework at all. My simple loop publishing a message to a web client eats _loads_ of CPU managing only 1000 messages/sek. All with loglevel disabled to minimum. All time is spent in LibLog which calls reflection and Expression.Compile() on _every_ published message.

**Comment by darkl** [comment]
LibLog is designed for libraries. The only places in the code where `.Compile()` is called (at least for the Serilog provider) is in static context (i.e. static constructor or the SerilogLogProvider constructor which happens once). This should be ok for most cases.

**Comment by darkl** [comment]
I also just submitted [a pull request](https://github.com/damianh/LibLog/pull/141/files) which fixes all Expression.Compile per instance/per message calls to static calls. This should resolve all such issues if they exist.

**Comment by llehn** [comment]
Thank you! I just tried your changes with WampSharp - throughput went from 1200 messages/sek to 13000 messages/sek. So the issue did exist. 
I'll do a PR to this repo using the code from your PR.
